### ISSUE TITLE
Inconsistency in when a send should be enabled between RDMC and SMC

### ISSUE THREAD

**Issue by songweijia** [Derecho-Project/derecho#175]
Aahil and Peter evaluated the performance of typed_subgroup_bw_test with persistence enabled. They found that SMC is many times faster than RDMC with the same message size and window size settings. I verified it using 10KB messages.

After digging into the code, I found that RDMC send and SMC send fall into quite different path even before calling `rdmc::send()` or `sst:multicast_group::send()`. Most importantly, the semantics of them are different.

On the RDMC path, the sending task is managed by the send loop thread defined in `derecho::MulticastGroup::send_loop()`. It relies on a condition `should_send()` to decide when to send the next message. The `should_send` condition requires that the next message to send should not exceed the sequence of lastest delivered or persisted message plus window size. However, on the SMC path, it does not test the persisted message sequence number. 

This different prevent deep batching for persistence on RDMC path, resulting in low performance with persistence+RDMC.

The code was about 4 years old. And I'm wondering, why the send care about lasted persisted message? Can we just test the latest delivered number on the RDMC path?

Please see the code [here](
https://github.com/Derecho-Project/derecho/blob/01a08f481959bff45f61086590b8b87e01fa003d/src/core/multicast_group.cpp#L1011).

**Comment by KenBirman** [comment]
I’m not really convinced that there is an issue here.  RDMC is designed only for messages that can be pipelined as a series of large chunks, such as a photo or video.  Because of the chunking, we send as a series of actions and need to finish one message before we can start to send the next.

So both the performance and the code you mentioned make sense.  The performance is low because you are giving RDMC a worst case situation: a small object far below the pipeline chunking threshold where it can run as a steady data flow.  And the sequence check is to ensure that we aren’t sending two messages simultaneously, because RDMC is totally unable to deal with such a case.  

SMC in contrast is designed for objects smaller than the pipelines chunking opportunity, and because it copies the object into the SST, doesn’t need to worry about a disordered chunk delivery that would violate the hard-wired binomial pipeline rule, which is very specific about what chunk arrives in what order at each node within the RDMC group.  SMC is getting this from the SST slots in the round robin message buffer it uses.

So, basically, short of doing an additional transport protocol for middle sized messages that are too big for SMC but too small for RDMC to work efficiently, I’m not seeing what we could do here.  A whole new transport for middle sized multicast could be a sensible thing to try, but would be a huge amount of work.  I’m not saying it is a bad idea, but just that for me, it isn’t an obvious priority right now!

**Comment by songweijia** [comment]
I agree that RDMC is for the bigger message. I'm suggesting a very simple change: allows the RDMC continues to send messages as long as we don't exceed the latest delivered message plus window size.This will not break RDMC but give RDMC more oppurtunity to batch up for short messages. And SMC and RDMC semantics will be aligned.

It means just removing the second condition in the `if` test below:
```
            for(uint i = 0; i < num_shard_members; ++i) {
                if(sst->delivered_num[node_id_to_sst_index.at(shard_members[i])][subgroup_num] < static_cast<message_id_t>((msg.index - subgroup_settings.profile.window_size) * num_shard_senders + shard_sender_index)
                   || (sst->persisted_num[node_id_to_sst_index.at(shard_members[i])][subgroup_num] < static_cast<message_id_t>((msg.index - subgroup_settings.profile.window_size) * num_shard_senders + shard_sender_index))) {
                    return false;
                }
            }
```

**Comment by KenBirman** [comment]
I’m not at all convinced that this change would be correct.  Anyhow, RDMC never batches.  It would be a great addition to the system to have it batch, and very much in tune with Sagar and Lorenzo’s Spindle techniques.  On the other hand it may be very hard to do.

To batch RDMC messages, which are normally sent in a zero copy way, we would have to have some form of scatter gather RDMA send, plus send the various messages sizes for the batch members.  The sitting code for receiving would be fine, I think, but would have to do each of the inner delivery upcalls one by one.

So I am not opposed to the idea and I agree that it could be a win (or a MMC protocol, for that matter, could be a win).  But I just think this is a harder thing than just modifying that line.  I understand what you are saying but if we remove that line, I’m very worried that messages could end up sent concurrently without any mechanism to keep track of which chunk is from which message.  And that layer of our code is phenomenally subtle.  Only Sagar could possibly make such a change...

**Comment by songweijia** [comment]
OK, I guess there is one possible misunderstanding between us I:

By "batching", I mean batching in the persistence thread (NOT batching in RDMC data path like what Sagar and Lorenzo's work). The batching I'm talking about is, writing multiple messages into persistent layer together, which will save IO flushes. And hence improving the end-to-end performance when **persistence is enabled**.

What I didn't understand is, before we send an RDMC message, we need to wait on the persistent progress as the RDMC buffer has been released on delivery.

**Comment by KenBirman** [comment]
Ok, this makes more sense to me now.  In fact I still think that part of the story here is that in the sender side, a middle sized message won’t be handled very efficiently.

On the receive side, we actually do have a way to detect that a batch of messages are deliverable, and we could try to hand off batches of messages... but only if they come as a batch in SMC.  With RDMC you are correct, we lack a way to do that.  Is this really limiting performance, though?  Even without such a batches delivery for RDMC, we are still running many x faster than the data transfer rate to a flash disk or Optane, aren’t we?  I think it was a factor of 10x or more last time we looked closely at that.

**Comment by songweijia** [comment]
I think the medium message size is fine. When RDMC performance begins to drop as message size goes smaller, the SMC can take over. The "divider" point might change with the group size, number of senders, and the network bandwidth. Choosing an optimal divider can almost achieve the peak throughput. So, I'm not worry about RDMC performance without batching in its own data path.

@sagarjha You might want to confirm the change I want to made to your code [here](https://github.com/Derecho-Project/derecho/blob/01a08f481959bff45f61086590b8b87e01fa003d/src/core/multicast_group.cpp#L1011).

**Comment by songweijia** [comment]
To be rewritten.

**Comment by sagarjha** [comment]
Hi @songweijia, it simply looks like a bug.

SST should make sure that a message has been persisted before overwriting the slot with a new message. So we should be checking the persisted_num value as we do in RDMC send_loop.

Does that make sense?

**Comment by songweijia** [comment]
@sagarjha, I believe SST is doing the right thing: once a message is delivered, the buffer can be ready for new incoming messages. Persistence thread is reading from the delivered state, instead of the incoming message buffer. So, what needs to be changed is the RDMC data path behavior.

**Comment by sagarjha** [comment]
I see. It's not the messages being persisted, but the object state. Makes sense.

**Comment by songweijia** [comment]
Should we close this one and move to #176?

**Comment by sagarjha** [comment]
sure.
